### ISSUE TITLE
AD1939 driver fails to compile against 6.1.38 LTS kernel

### ISSUE THREAD

**Issue by LRitzdorf** [ADSD-SoC-FPGA/Code#5]
The AD1939 driver provided here does not compile against kernel 6.1.38 (the latest LTS branch from [Altera's repo](https://github.com/altera-opensource/linux-socfpga)). Specifically, it complains of the following:
```
linux/ad1939/ad1939.c:185:14: error: implicit declaration of function ‘spi_busnum_to_master’; did you mean ‘spi_alloc_master’? [-Werror=implicit-function-declaration]
     master = spi_busnum_to_master( spi_device_info.bus_num );
              ^~~~~~~~~~~~~~~~~~~~
              spi_alloc_master
```
Based on some quick research, it [appears](https://forums.raspberrypi.com/viewtopic.php?t=352916#p2115487) that the `spi_busnum_to_master` function was previously removed from the kernel.

Relatedly, it is worth noting that Analog Devices has contributed [their own driver](https://wiki.analog.com/resources/tools-software/linux-drivers/sound/ad1936) to the Linux kernel. While I realize that the custom drivers in this repo may perform certain special initialization tasks, this upstreamed driver may at least be useful as a reference for "the right way" to implement such code.

Further information on the native driver is available on the Analog Devices wiki for the AD1939, particularly regarding the [specification of relevant device settings](https://wiki.analog.com/resources/tools-software/linux-drivers/sound/ad1936#declaring_spi_slave_devices) (which must apparently be hard-coded in a device-specific file).

\
Related: #4. Note, however, that this issue relates to an actual compilation-stopping error, not simply a warning.

**Comment by tvannoy** [comment]
> Relatedly, it is worth noting that Analog Devices has contributed [their own driver](https://wiki.analog.com/resources/tools-software/linux-drivers/sound/ad1936) to the Linux kernel. While I realize that the custom drivers in this repo may perform certain special initialization tasks, this upstreamed driver may at least be useful as a reference for "the right way" to implement such code.

That driver has been around since 2010, and is certainly the more correct way to do this. We had [used the official driver for one project](https://github.com/fpga-open-speech-tools/component_library/tree/8b5d1f81a77ed59ed249acb3159c797e03203206/sodia_ip/ad193x) which made the [audio mini show up as sound card](https://github.com/fpga-open-speech-tools/de10nano_projects/tree/dev/AudioMini_Simple_Sound_Card)

I was not involved in that project, but I recall lots of grumbling about how difficult it was to set up the ad193x driver.

There is also a [mainline driver for the TPA headphone amplifier](https://www.ti.com/tool/TPA6130A2SW-LINUX) as well.

These drivers are definitely the correct way to do all this. They use the [ASoC subsystem](https://elinux.org/images/b/b5/Belloni-alsa-asoc.pdf). In theory, we could configure both the codec and headphone amp (almost?) entirely using the device tree. We spent a little bit of time looking at that, but never figured it out.

> the custom drivers in this repo may perform certain special initialization tasks

I don't think they do anything special that would cause us to not use the mainline drivers :slightly_smiling_face:. In fact, they don't implement many things at all :upside_down_face: 

> Based on some quick research, it [appears](https://forums.raspberrypi.com/viewtopic.php?t=352916#p2115487) that the spi_busnum_to_master function was previously removed from the kernel.

This is typical. Something similar happened with the i2c subsystem, which [affected our tpa613a2 driver](https://github.com/fpga-open-speech-tools/component_library/issues/22).

As another note, we should be putting these devices as children in the spi and i2c device tree nodes. Combine that with using the mainline drivers, and we'd be doing everything correctly and be golden :medal_sports: 

---

## Thoughts on fixing `spi_busnum_to_master`

As for the `spi_busnum_to_master` thing, here's what I'm finding so far.

[`spi_busnum_to_maser` was intended for non-static configurations](https://www.kernel.org/doc/Documentation/spi/spi-summary), which certainly isn't our situation:
> NON-STATIC CONFIGURATIONS
> 
> Developer boards often play by different rules than product boards, and one
> example is the potential need to hotplug SPI devices and/or controllers.
> 
> *For those cases you might need to use spi_busnum_to_master() to look
> up the spi bus master*, and will likely need spi_new_device() to provide the
> board info based on the board that was hotplugged.  Of course, you'd later
> call at least spi_unregister_device() when that board is removed.> 
>
> When Linux includes support for MMC/SD/SDIO/DataFlash cards through SPI, those
> configurations will also be dynamic.  Fortunately, such devices all support
> basic device identification probes, so they should hotplug normally.

The [Declare Slave Devices section of the spi overview](https://www.kernel.org/doc/html/latest/spi/spi-summary.html?highlight=spi#declare-slave-devices) suggests to me that we might just need to call `spi_register_board_info()` instead of `spi_busnum_to_master()` and `spi_new_device()`.

The description of `spi_new_device()` also suggests that we should be using `spi_register_board_info()` instead:
>  Some development platforms may not be able to use spi_register_board_info though, and this is exported so that for example a USB or parport based adapter driver could add devices (which it would learn about out-of-band).

If I am understanding the [spi api documentation](https://docs.kernel.org/driver-api/spi.html) correctly, I think we can just use `spi_register_board_info()` after creating the `spi_device_info` struct. Here's the description of `spi_register_board_info()`:
> Board-specific early init code calls this (probably during arch_initcall) with segments of the SPI device table. Any device nodes are created later, *after the relevant parent SPI controller (bus_num) is defined*. We keep this table of devices forever, so that reloading a controller driver will not make Linux forget about these hard-wired devices.

We know which bus number our device is on, and the device tree should have info that maps that bus number to the appropriate spi controller. 

My guess is that the code will become
```c
// Register the device
    struct spi_board_info spi_device_info = {
        .modalias = "al_ad1939_",     // Needs to be the same as the device name
        .max_speed_hz = speed,
        .bus_num = 1,             // Determined from /sys/class/spi_master                
        .chip_select = 0,                     // check
        .mode = 3,
    };
    
    ret = spi_register_board_info(spi_device_info, ARRAY_SIZE(spi_device_info));

    if (ret < 0)
    {
        # report error
    }
```

instead of
```c
// Register the device
    struct spi_board_info spi_device_info = {
        .modalias = "al_ad1939_",     // Needs to be the same as the device name
        .max_speed_hz = speed,
        .bus_num = 1,             // Determined from /sys/class/spi_master                
        .chip_select = 0,                     // check
        .mode = 3,
    };
    
    /*To send data we have to know what spi port/pins should be used. This information 
      can be found in the device-tree. */
    master = spi_busnum_to_master( spi_device_info.bus_num );
    if( !master ){
        printk("MASTER not found.\n");
            return -ENODEV;
    }
     
    // create a new slave device, given the master and device info
    spi_device = spi_new_device( master, &spi_device_info );
```

Please try this out and let me know how it goes. If needed, I can play around with this tomorrow and try to get the driver working with the 6.1.38 lts kernel.

**Comment by tvannoy** [comment]
@rksnider @w-culhane

**Comment by tvannoy** [comment]
Just FYI, it is not uncommon for internal APIs to change (see the [stable api nonsense](https://github.com/torvalds/linux/blob/master/Documentation/process/stable-api-nonsense.rst) file). That's one reason mainlining drivers is a good idea: you don't have to keep up with api changes because the subsystem maintainers do that.

@rksnider This situation begs the question of if we should suggest, in the textbook, to use specific versions of the LTS kernel to avoid future issues, at the risk of being completely out of date and not supported; or if we just continue to deal with these issues if they occasionally pop up.

**Comment by LRitzdorf** [comment]
> I recall lots of grumbling about how difficult setting up the ad193x driver was.
> These drivers are definitely the correct way to do all this. They use the [ASoC subsystem](https://elinux.org/images/b/b5/Belloni-alsa-asoc.pdf). In theory, we could configure both the codec and headphone amp (almost?) entirely using the device tree. We spent a little bit of time looking at that, but never figured it out.

Yep, that sounds about right. 🙂 Just wanted to point out the mainline drivers, in case they hadn't been considered before.

> As another note, we should be putting these devices as children in the spi and i2c device tree nodes.

Okay, this makes more sense. I was actually planning to ask about that later.
Would doing this affect the driver code further? I don't know how dependent the current implementation is on the precise device tree structure.

> Please try this out and let me know how it goes. If needed, I can play around with this tomorrow and try to get the driver working with the 6.1.38 lts kernel.

Thanks! Will have updates shortly.

**Comment by LRitzdorf** [comment]
Okay, question time, @tvannoy.

As a general observation, it seems like there are two flows we could follow here:
- Register a device as if we were in board-specific early init code
  - Uses `spi_register_board_info()`
  - Unclear how we  get a `struct spi_device *` after this (needed so we can actually send commands later)
- Register a new device "dynamically," as if we just learned about the device (via a hotplug event or similar)
  - Uses [`spi_new_device()`](https://docs.kernel.org/driver-api/spi.html#c.spi_new_device)
  - Unclear how we get the `struct spi_controller *` needed for this, now that `spi_busnum_to_master()` is gone
    - There's a [`__spi_alloc_controller()`](https://docs.kernel.org/driver-api/spi.html#c.__spi_alloc_controller) function call that would let us create our own controller; is that what we want?

**Comment by tvannoy** [comment]
> Would doing this affect the driver code further? I don't know how dependent the current implementation is on the precise device tree structure.

All that is in the current device tree nodes is the compatible string, so I wouldn't think moving the nodes from the root into the spi/i2c nodes would be an issue. IIRC, I did try that at some point, and I don't believe I had working audio, but I don't recall exactly what I did.

I think it's worth a try, once the drivers actually work again with the new kernel :slightly_smiling_face:

**Comment by tvannoy** [comment]
> that would allow us to have a single function for volume control, instead of one for each channel

IIRC, @rksnider's general philosophy has been having separate controls for each channel (one could argue that there generally isn't much need to have separate channel controls, which I generally agree with).

**Comment by tvannoy** [comment]
### Register a device as if we were in board-specific early init code 
> - Unclear how we get a struct spi_device * after this (needed so we can actually send commands later)

I also don't see any obvious way (or any way at all) to to get a `spi_device` struct in this approach either. My bad...

---

### Register a new device "dynamically," as if we just learned about the device (via a hotplug event or similar) 

> Unclear how we get the struct spi_controller * needed for this, now that spi_busnum_to_master() is gone

I also don't see any way to get a `spi_controller` either...

I don't believe we want register our own controller, as this driver is a "protocol driver" not a "controller driver" (from the spi API):
> The programming interface is structured around two kinds of driver, and two kinds of device. A "Controller Driver" abstracts the controller hardware, which may be as simple as a set of GPIO pins or as complex as a pair of FIFOs connected to dual DMA engines on the other side of the SPI shift register (maximizing throughput). Such drivers bridge between whatever bus they sit on (often the platform bus) and SPI, and expose the SPI side of their device as a [struct spi_controller](https://docs.kernel.org/driver-api/spi.html#c.spi_controller). SPI devices are children of that master, represented as a [struct spi_device](https://docs.kernel.org/driver-api/spi.html#c.spi_device) and manufactured from [struct spi_board_info](https://docs.kernel.org/driver-api/spi.html#c.spi_board_info) descriptors which are usually provided by board-specific initialization code. A [struct spi_driver](https://docs.kernel.org/driver-api/spi.html#c.spi_driver) is called a "Protocol Driver", and is bound to a spi_device using normal driver model calls.

Moreover, the spi controller must already exist because there is a [device tree node](https://elixir.bootlin.com/linux/v6.1.38/source/arch/arm/boot/dts/socfpga.dtsi#L829) for it, and we have it enabled (`spi0`, in our case). I suppose we could determine if a spi controller driver is actually loaded via `lsmod` and probing around `/sys/`.

Perhaps an easy solution is to just [copy what `spi_busnum_to_master` used to do](https://android.googlesource.com/kernel/msm/+/android-msm-vega-4.4-oreo-daydream/drivers/spi/spi.c#1980):
```c
struct spi_master *spi_busnum_to_master(u16 bus_num)
{
	struct device		*dev;
	struct spi_master	*master = NULL;
	dev = class_find_device(&spi_master_class, NULL, &bus_num,
				__spi_master_match);
	if (dev)
		master = container_of(dev, struct spi_master, dev);
	/* reference got in class_find_device */
	return master;
}
```

For slightly more context, I did find some of the commits where these changes happened:
- https://github.com/altera-opensource/linux-socfpga/commit/8caab75fd2c2a92667cbb1cd315720bede3feaa9
- https://github.com/altera-opensource/linux-socfpga/commit/bdc7ca008e1f5539e891187032cb2cbbc3decb5e
- [relevant kernel mailing list](https://lore.kernel.org/all/20211007121415.2401638-1-u.kleine-koenig@pengutronix.de/)

If we go this route, we should change `spi_master` to `spi_controller`, per some of the above commits that renamed `spi_master` to `spi_controller`.

---

### Add the AD1939 to the device tree as a child of the SPI bus, and write a more "proper" driver that binds to it

This is definitely a more preferable solution, but would likely take longer to implement. This is probably less work than figuring out the mainline ASoC drivers (which would tie into ALSA, which would obviously be very cool, but also not 100% necessary for us).

> Currently, this seems to be just init'ing registers and facilitating volume control/sample frequency; is there any other key functionality I'm missing?

I don't think you're missing anything. All it does, as far as I remember, is set up some initial settings and provide sysfs attributes for the registers we care about.

---

### Use an older kernel :scream: 

Our last option, which is actually the easiest of all, but not very satisfying and not a long-term solution, is to just use an older kernel where `spi_busnum_to_master` still exists. Thus we just kick this can down the road...

Now that we know which commit removed `spi_busnum_to_master`, we could determine the newest lts release before that change.

**Comment by LRitzdorf** [comment]
> > that would allow us to have a single function for volume control, instead of one for each channel
> 
> IIRC, @rksnider's general philosophy has been having separate controls for each channel (one could argue that there generally isn't much need to have separate channel controls, which I generally agree with).

I didn't phrase that very well, sorry. I agree with having separate controls, but there's no need to have a separate *function call* in the code for each channel. I've found a way of adding extra device attribute data that's visible to (for example) a volume-update function, and which changes based on where that function is called from. It's kind of a mess of kernel macros, but structurally looks like this:
- Custom wrapper struct
  - Normal kernel device attribute struct
    - sysfs filename and permissions
    - show and store function callbacks
  - Extra data (e.g. SPI address for relevant volume channel)

Normally, the show/store functions are unique for each register. But with this system, they can be the same function, which will then be able to consult the extra data in our custom wrapper struct. This allows us to externalize the only thing that changes between the different volume channels: the SPI register address with which to interact.

Anyway, that's a problem for a different PR, if any 🙂

**Comment by LRitzdorf** [comment]
Okay — after a bunch of reading through kernel code (specifically [`spidev.c`](https://elixir.bootlin.com/linux/v6.1.38/source/drivers/spi/spidev.c), a generic implementation that may be very useful here, I think I understand the driver's interaction with the device tree a bit better now.

The driver's `*_probe` function is passed a `struct spi_device *` when it's called by... uhh, other kernel machinery? The `spidev` driver does some fancy stuff where it uses a file struct's "misc private data" pointer to keep track of details about the device struct, but I think that's because it needs to potentially handle a bunch of different SPI devices at once. We should just be able to save that data into... a global variable or something, I guess?

The key point here is that, if we place the AD1939 device tree node under the SPI bus node, I know how to get an already-valid `struct spi_device *` for it!
This makes "write a more proper driver" approach much more feasible. We'd just have to implement those key features from earlier: init'ing the codec, and exposing key registers in the sysfs.

That seems completely doable to me; the only possible difficulty I see is figuring out exactly which attributes need to be specified in the device tree ([max speed, chip select, bits per word...](https://github.com/ADSD-SoC-FPGA/Code/blob/85a98c7ce6ea5d8074de3a742dcdcb41fbc1a6af/examples/passthrough/linux/ad1939/ad1939.c#L176-L201)). Some quick searching found [this SO post](https://stackoverflow.com/q/52345615), which I'm linking here mostly as a reference for future-me. There are probably better ones out there, too.

Also, for other students, downgrading the kernel is probably the easiest short-term solution, unsatifying as it is 😐

**Comment by tvannoy** [comment]
> > > that would allow us to have a single function for volume control, instead of one for each channel
> > 
> > 
> > IIRC, @rksnider's general philosophy has been having separate controls for each channel (one could argue that there generally isn't much need to have separate channel controls, which I generally agree with).
> 
> I didn't phrase that very well, sorry. I agree with having separate controls, but there's no need to have a separate _function call_ in the code for each channel. I've found a way of adding extra device attribute data that's visible to (for example) a volume-update function, and which changes based on where that function is called from. It's kind of a mess of kernel macros, but structurally looks like this:
> 
>     * Custom wrapper struct
>       
>       * Normal kernel device attribute struct
>         
>         * sysfs filename and permissions
>         * show and store function callbacks
>       * Extra data (e.g. SPI address for relevant volume channel)
> 
> 
> Normally, the show/store functions are unique for each register. But with this system, they can be the same function, which will then be able to consult the extra data in our custom wrapper struct. This allows us to externalize the only thing that changes between the different volume channels: the SPI register address with which to interact.
> 
> Anyway, that's a problem for a different PR, if any 🙂

I personally have no problem with the simplicity of having a show/store function for each attribute, and it appears that plenty of other drivers do the same thing. Of course this can lead to a lot of duplication, especially in our case.

If you don't like having all the function prototype definitions, most drivers I've seen don't bother with prototypes and just define the functions in the correct order.

Certainly there is utility for having a single show/store function when all the attributes behave exactly the same. I think there is less utility if the attributes are all slightly different (e.g., different data types), because then you end up with a case statement or similar inside the show/store function.

You're approach was probably something similar to using a [`dev_ext_attribute`](https://elixir.bootlin.com/linux/latest/source/include/linux/device.h#L119) struct, which allows you to tack on some extra data which will get passed into the show/store function via the `device_attr *`. Like you said, the obvious use case for this is when the only thing that is different is something like the address of register offset, in which case no if/else logic is required in the show/store function. I can see this being helpful, for example, when all registers in the FPGA IP core are the same exact data type (from both the userspace and hardware perspective) and there is no need for any special handling. When userspace has different expectations for what data type the registers are (e.g., an `enable` register and a `data` register), then there would probably need to be additional logic in the unified show/store function to handle those data type differences.

In an educational context, I'd prefer keeping things simple for new learners to understand, even if that sometimes results in evil copy/paste duplication.

But yes, as you mentioned, this is a different problem entirely. It would be satisfying to have less duplication.

**Comment by tvannoy** [comment]
I was hopeful that we might be able to use [`of_find_spi_controller_by_node()`](https://elixir.bootlin.com/linux/v6.1.38/source/drivers/spi/spi.c#L4341), since it does something similar to `spi_busnum_to_master`; in theory, we'd just need the ad1939 dt node to be a child of the spi controller, then we could get the spi controller's dt node. However, that function is static (private), as are all of the structs and things we would need to recreate that function or `spi_busnum_to_master`. For what it's worth, the code that is in the `init` function right now really should go into the `probe` function instead.

So it looks like we are down to a few options
1. Using an old kernel for now
2. Getting the ad1939 to behave as a spi slave in the dt, and modifying the driver accordingly (all we really need to do is change how we get the `spi_device`---easier said than done, of course).
3. [Use `spidev` in userspace](https://wiki.st.com/stm32mpu/wiki/How_to_use_SPI_from_Linux_userland_with_spidev). This is maybe a bit painful, but we do only set a few register values once during initialization. It's rare that we ever change anything later.

I don't think we should spend too much time recreating and cleaning up the current driver. A slightly better "stop-gap" driver is a still a stop-gap between our current situation and using the mainline drivers with ASoC and ALSA. 

I'm seeing if I can hack something together today. If we can't get something working by tonight, given the time-sensitivity of this with respect to the class/lab, then we should just move forward for this semester using an older kernel version.

**Comment by tvannoy** [comment]
> I didn't know there was a built-in implementation

Neither did I until today, but I recalled seeing that having one show/store for multiple attributes was possible.

**Comment by tvannoy** [comment]
> The driver's *_probe function is passed a struct spi_device * when it's called by... uhh, other kernel machinery?

That seems about right. Kernel (plus device tree) machinery magically gave me a `spi_device *` and everything worked trivially. Hoorah! :tada:

**Comment by LRitzdorf** [comment]
Nice!

On the one hand, yes, mainline drivers would be great. On the other, it almost feels easier for students to just be presented with the register files they're used to, instead of *also* having to learn about Linux audio systems. Also, I would think that a "proper" device tree-based driver would be pretty maintainable, since the device tree infrastructure shouldn't disappear anytime remotely soon... Food for thought, and conversation with @rksnider, I suppose.

**Comment by tvannoy** [comment]
In terms of using the mainline audio drivers, everything would be integrated with ALSA, so they'd just use alsamixer (or similar) to adjust the volume if needed (in practice, we never really touch the registers for the codec and headphone amp except for initialization).

> a "proper" device tree-based driver would be pretty maintainable, since the device tree infrastructure shouldn't disappear anytime remotely soon

Yes, using the device tree and the i2c and spi subsystems as intended would be easier to maintain and should lead to less issues, since generally it's the internal APIs that change. I wouldn't exactly call device tree bindings stable, though; in the very least, the documentation is in flux, and there is talk of moving device trees out of the linux kernel source tree entirely.

**Comment by tvannoy** [comment]
Resolved by #7
